tekton/integration_pipeline.go

```diff
@@ -0,0 +1,167 @@
+	return &r.PipelineRun
+}
+
+// NewIntegrationPipelineRun creates an empty PipelineRun in the given namespace. The name will be autogenerated,
+// using the prefix passed as an argument to the function.
+func NewIntegrationPipelineRun(prefix, namespace string, integrationTestScenario v1alpha1.IntegrationTestScenario) *IntegrationPipelineRun {
+	pipelineRun := tektonv1beta1.PipelineRun{
+		ObjectMeta: v1.ObjectMeta{
+			GenerateName: prefix + "-",
+			Namespace:    namespace,
+		},
+		Spec: tektonv1beta1.PipelineRunSpec{
+			PipelineRef: &tektonv1beta1.PipelineRef{
+				Name:   integrationTestScenario.Spec.Pipeline,
+				Bundle: integrationTestScenario.Spec.Bundle,
+			},
+		},
+	}
+	return &IntegrationPipelineRun{pipelineRun}
+}
+
+// WithSpecialParams adds Pipeline reference, params to the PipelineRun.
+func (r *IntegrationPipelineRun) WithSpecialParams(integration *v1alpha1.IntegrationTestScenario) *IntegrationPipelineRun {
+	r.Spec.PipelineRef = &tektonv1beta1.PipelineRef{
+		Name:   integration.Spec.Pipeline,
+		Bundle: integration.Spec.Bundle,
+	}
+
+	valueType := tektonv1beta1.ParamTypeString
+
+	for _, param := range integration.Spec.Params {
+		if len(param.Value) > 0 {
+			valueType = tektonv1beta1.ParamTypeArray
+		}
+
+		r.WithExtraParam(param.Name, tektonv1beta1.ArrayOrString{
+			Type:      valueType,
+			StringVal: param.Value,
+			ArrayVal:  param.Values,
+		})
+	}
+
+	return r
+}
+
+// WithExtraParam adds an extra param to the Integration PipelineRun. If the parameter is not part of the Pipeline
+// definition, it will be silently ignored.
+func (r *IntegrationPipelineRun) WithExtraParam(name string, value tektonv1beta1.ArrayOrString) *IntegrationPipelineRun {
+	r.Spec.Params = append(r.Spec.Params, tektonv1beta1.Param{
+		Name:  name,
+		Value: value,
+	})
+
+	return r
+}
+
+// WithApplicationSnapshot adds a param containing the ApplicationSnapshot as a json string
+// to the integration PipelineRun.
+func (r *IntegrationPipelineRun) WithApplicationSnapshot(snapshot *appstudioshared.ApplicationSnapshot) *IntegrationPipelineRun {
+	// We ignore the error here because none should be raised when marshalling the spec of a CRD.
+	// If we end up deciding it is useful, we will need to pass the errors trough the chain and
+	// add something like a `Complete` function that returns the final object and error.
+	snapshotString, _ := json.Marshal(snapshot.Spec)
+
+	r.WithExtraParam("SNAPSHOT", tektonv1beta1.ArrayOrString{
+		Type:      tektonv1beta1.ParamTypeString,
+		StringVal: string(snapshotString),
+	})
+
+	if r.ObjectMeta.Labels == nil {
+		r.ObjectMeta.Labels = map[string]string{}
+	}
+	r.ObjectMeta.Labels[SnapshotNameLabel] = snapshot.Name
+
+	return r
+}
+
+// WithIntegrationLabels adds the type and IntegrationTestScenario name as labels to the Integration PipelineRun.
+func (r *IntegrationPipelineRun) WithIntegrationLabels(scenarioName string) *IntegrationPipelineRun {
```

Comment:
I think we can have a full `integrationTestScenario` object as the function's parameter here, instead of just the `scenarioName` string, this will match the other `With*` functions.